hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
+{-# LANGUAGE TypeApplications    #-}
+
+{-|
+Description : Allowed effects for interacting with Nar files.
+Maintainer  : Shea Levy <shea@shealevy.com>
+|-}
+module System.Nix.Nar where
+
+import           Control.Applicative
+import           Control.Monad              (replicateM, replicateM_)
+import qualified Data.Binary                as B
+import qualified Data.Binary.Get            as B
+import qualified Data.Binary.Put            as B
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import           Data.Maybe                 (fromMaybe)
+import           Data.Monoid                ((<>))
+import qualified Data.Set                   as Set
+import qualified Data.Text                  as T
+import qualified Data.Text.Encoding         as E
+import           GHC.Int                    (Int64)
+
+import           System.Nix.Path
+
+
+data NarEffects (m :: * -> *) = NarEffets {
+    readFile         :: FilePath -> m BSL.ByteString
+  , listDir          :: FilePath -> m [FileSystemObject]
+  , narFromFileBytes :: BSL.ByteString -> m Nar
+  , narFromDirectory :: FilePath -> m Nar
+}
+
+
+-- Directly taken from Eelco thesis
+-- https://nixos.org/%7Eeelco/pubs/phd-thesis.pdf
+
+-- TODO: Should we use rootedPath, validPath rather than FilePath?
+data Nar = Nar { narFile :: FileSystemObject }
+    deriving (Eq, Ord, Show)
+
+data FileSystemObject =
+    Regular IsExecutable BSL.ByteString
+  | Directory (Set.Set (PathName, FileSystemObject))
+  | SymLink BSL.ByteString
+  deriving (Eq, Show)
+
+-- TODO - is this right? How does thesis define ordering of FSOs?
+instance Ord FileSystemObject where
+    compare (Regular _ c1) (Regular _ c2) = compare c1 c2
+    compare (Regular _ _)  _              = GT
+    compare (Directory s1) (Directory s2) = compare s1 s2
+    compare (Directory _)  _              = GT
+    compare (SymLink l1) (SymLink l2)     = compare l1 l2
+
+data IsExecutable = NonExecutable | Executable
+    deriving (Eq, Show)
+
+
+------------------------------------------------------------------------------
+-- | Serialize Nar to lazy ByteString
+putNar :: Nar -> B.Put
```

Comment:
Hmm... For this to be efficient our nar would have to be constructed with unsafeInterleaveIO. Not necessarily a problem, but I'm wondering if that's the API we want to expose. Maybe it's the simplest for now.

Reply:
Since we are producing a lazy bytestring, can't `putNar` run in constant space? Only if we forced the `Nar` would I expect this approach to blow up. Well, my intuition here is completely un-tested :) so...

Reply:
To have a `Nar` in the first place requires us to have already recursed all the directories and read all the files, unless we're interleaving IO, right? How else would laziness save us here?